packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -122,928 +126,785 @@ function attributeExpressionReferencesForEachIndex(
         }
+
+        parsedTextNodes.push(createText(node, value));
     }
 
-    function applyHandlers(element: IRElement, parsedAttr: ParsedAttribute) {
-        let eventHandlerAttribute = parsedAttr.pick(EVENT_HANDLER_RE);
-        while (eventHandlerAttribute) {
-            if (eventHandlerAttribute.type !== IRAttributeType.Expression) {
-                return warnOnIRNode(
-                    ParserDiagnostics.EVENT_HANDLER_SHOULD_BE_EXPRESSION,
-                    eventHandlerAttribute
-                );
-            }
+    return parsedTextNodes;
+}
 
-            let eventName = eventHandlerAttribute.name;
-            if (!eventName.match(EVENT_HANDLER_NAME_RE)) {
-                return warnOnIRNode(ParserDiagnostics.INVALID_EVENT_NAME, eventHandlerAttribute, [
-                    eventName,
-                ]);
-            }
+function parseComment(node: parse5.CommentNode): IRComment {
+    const value = decodeTextContent(node.data);
+    return createComment(node, value);
+}
+
+function getTemplateRoot(
+    ctx: ParserCtx,
+    documentFragment: parse5.DocumentFragment
+): parse5.Element {
+    // Filter all the empty text nodes
+    const validRoots = documentFragment.childNodes.filter(
+        (child) =>
+            parse5Utils.isElementNode(child) ||
+            (parse5Utils.isTextNode(child) && child.value.trim().length)
+    );
 
-            // Strip the `on` prefix from the event handler name
-            eventName = eventHandlerAttribute.name.slice(2);
+    if (validRoots.length > 1) {
+        const duplicateRoot = validRoots[1].sourceCodeLocation!;
+        ctx.throwAtLocation(ParserDiagnostics.MULTIPLE_ROOTS_FOUND, duplicateRoot);
+    }
 
-            const on = element.on || (element.on = {});
-            on[eventName] = eventHandlerAttribute.value;
+    const [root] = validRoots;
 
-            eventHandlerAttribute = parsedAttr.pick(EVENT_HANDLER_RE);
-        }
+    if (!root || !parse5Utils.isElementNode(root)) {
+        ctx.throwAtLocation(ParserDiagnostics.MISSING_ROOT_TEMPLATE_TAG);
     }
 
-    function applyIf(element: IRElement, parsedAttr: ParsedAttribute) {
-        const ifAttribute = parsedAttr.pick(IF_RE);
-        if (ifAttribute) {
-            if (ifAttribute.type !== IRAttributeType.Expression) {
-                return warnOnIRNode(
-                    ParserDiagnostics.IF_DIRECTIVE_SHOULD_BE_EXPRESSION,
-                    ifAttribute
-                );
-            }
+    return root;
+}
 
-            const [, modifier] = ifAttribute.name.split(':');
-            if (!VALID_IF_MODIFIER.has(modifier)) {
-                return warnOnIRNode(ParserDiagnostics.UNEXPECTED_IF_MODIFIER, ifAttribute, [
-                    modifier,
-                ]);
-            }
+function applyHandlers(ctx: ParserCtx, element: IRElement, parsedAttr: ParsedAttribute) {
+    let eventHandlerAttribute = parsedAttr.pick(EVENT_HANDLER_RE);
+    while (eventHandlerAttribute) {
```

Comment:
Unrelated to this PR, but it's something that I haven't caught before. This function contains duplicated `parsedAttr.pick(EVENT_HANDLER_RE)` before and at the end of the loop. It is preferable to have something like this to avoid the duplication:

```ts
let eventHandlerAttribute;
while (eventHandlerAttribute = parsedAttr.pick(EVENT_HANDLER_RE)) {
```

Reply:
I was looking at this one too, I'll add the update in the next PR.